paasta_tools/check_spark_jobs.py

```diff
@@ -0,0 +1,170 @@
+import argparse
+import datetime
+import logging
+from collections import defaultdict
+
+import requests
+
+from paasta_tools import mesos_tools
+from paasta_tools.monitoring_tools import send_event
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import list_services
+
+
+logger = logging.getLogger(__name__)
+
+
+JUPYTER_PREFIX = 'jupyterhub_'
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description='Reports long-running Spark frameworks.',
+    )
+    parser.add_argument(
+        '--min-hours',
+        type=float,
+        help='Report frameworks that have been registered for more than this duration',
+        default=0,
+    )
+    parser.add_argument(
+        '--no-notify',
+        action='store_true',
+        help='Skip notifying the teams that own each framework',
+    )
+    return parser.parse_args()
+
+
+def get_time_running(framework):
+    registered_time = datetime.datetime.fromtimestamp(framework['registered_time'])
+    return datetime.datetime.now() - registered_time
+
+
+def get_spark_properties(framework):
+    webui_url = framework.get('webui_url')
+    env_endpoint = f'{webui_url}/api/v1/applications/{framework.id}/environment'
+
+    try:
+        response = requests.get(env_endpoint, timeout=5)
+    except (requests.exceptions.Timeout, requests.exceptions.ConnectionError) as e:
+        logger.warning(f'Unable to connect to {env_endpoint}: {e!r}')
+        return None
+
+    if response.status_code != 200:
+        logger.warning(f'Bad response from {env_endpoint}: {response.status_code}')
+        return None
+
+    try:
+        return response.json()['sparkProperties']
+    except (ValueError, KeyError) as e:
+        logger.warning(f'Unable to get sparkProperties for {framework.id}: {e!r}')
+        return None
+
+
+def guess_service(properties):
+    if not properties:
+        return None
+    for key, value in properties:
+        if key == 'spark.executorEnv.PAASTA_SERVICE':
+            service = value
+            break
+    else:
+        return None
+    if service.startswith(JUPYTER_PREFIX):
+        return service[len(JUPYTER_PREFIX):]
+    else:
+        return service
+
+
+def get_matching_framework_info(min_hours):
+    frameworks = mesos_tools.get_all_frameworks(active_only=True)
+    matching_info = []
+    min_timedelta = datetime.timedelta(hours=min_hours)
+    for framework in frameworks:
+        if not framework.active:
+            continue
+        if framework.get('principal') != 'spark':
+            continue
+        time_running = get_time_running(framework)
+        if time_running >= min_timedelta:
+            info = {
```

Comment:
This function is good. I think it might be easier to understand and test if it *only* filtered long running spark frameworks from an input list of frameworks and literally returned the framework objects, and let some other function worry about extracting the pieces of data you need (service, url, etc).

But it does mean you would have to `get_time_running` more than once, which isn't super efficient. 

But in general, this function does 3 thing, fetches, filters, and creates a custom return object. Again I think this is just fine for this script, but was just a thought I had as I was reading the code.

Reply:
Yeah, I like the slight efficiency for now, but if the return output gets too much more complicated in the future, I'll separate it.